source/extensions/filters/http/cache/http_cache.h

```diff
@@ -0,0 +1,322 @@
+  // Prereq: first == UINT64_MAX || first <= last
+  // Invariant: isSuffix() || firstBytePos() <= lastBytePos
+  // Examples: RawByteRange(0,4) requests the first 5 bytes.
+  //           RawByteRange(UINT64_MAX,4) requests the last 4 bytes.
+  RawByteRange(uint64_t first, uint64_t last) : first_byte_pos_(first), last_byte_pos_(last) {
+    RELEASE_ASSERT(isSuffix() || first <= last, "Illegal byte range.");
+  }
+  bool isSuffix() const { return first_byte_pos_ == UINT64_MAX; }
+  uint64_t firstBytePos() const {
+    ASSERT(!isSuffix());
+    return first_byte_pos_;
+  }
+  uint64_t lastBytePos() const {
+    ASSERT(!isSuffix());
+    return last_byte_pos_;
+  }
+  uint64_t suffixLength() const {
+    ASSERT(isSuffix());
+    return last_byte_pos_;
+  }
+
+private:
+  uint64_t first_byte_pos_;
+  uint64_t last_byte_pos_;
+};
+
+// Byte range from an HTTP request, adjusted for a known response body size, and converted from an
+// HTTP-style closed interval to a C++ style half-open interval.
+class AdjustedByteRange {
+public:
+  // Construct an AdjustedByteRange representing the [first,last) bytes in the
+  // response body. Prereq: first <= last Invariant: begin() <= end()
+  // Example: AdjustedByteRange(0,4) represents the first 4 bytes.
+  AdjustedByteRange(uint64_t first, uint64_t last) : first_(first), last_(last) {
+    ASSERT(first < last, "Illegal byte range.");
+  }
+  uint64_t begin() const { return first_; }
+  // Unlike RawByteRange, end() is one past the index of the last offset.
+  uint64_t end() const { return last_; }
+  uint64_t length() const { return last_ - first_; }
+  void trimFront(uint64_t n) {
+    RELEASE_ASSERT(n <= length(), "Attempt to trim too much from range.");
+    first_ += n;
+  }
+
+private:
+  uint64_t first_;
+  uint64_t last_;
+};
+
+inline bool operator==(const AdjustedByteRange& lhs, const AdjustedByteRange& rhs) {
+  return lhs.begin() == rhs.begin() && lhs.end() == rhs.end();
+}
+
+std::ostream& operator<<(std::ostream& os, const AdjustedByteRange& range);
+
+// Adjusts request_range_spec to fit a cached response of size content_length, putting the results
+// in response_ranges. Returns true if response_ranges is satisfiable (empty is considered
+// satisfiable, as it denotes the entire body).
+// TODO(toddmgreer): Merge/reorder ranges where appropriate.
+bool adjustByteRangeSet(std::vector<AdjustedByteRange>& response_ranges,
+                        const std::vector<RawByteRange>& request_range_spec,
+                        uint64_t content_length);
+
+// Result of a lookup operation, including cached headers and information needed
+// to serve a response based on it, or to attempt to validate.
+struct LookupResult {
+  // If cache_entry_status_ == Unusable, none of the other members are
+  // meaningful.
+  CacheEntryStatus cache_entry_status_ = CacheEntryStatus::Unusable;
+
+  // Headers of the cached response.
+  Http::HeaderMapPtr headers_;
```

Comment:
This will require copying the headers for every cached response, which seems not-optimal. Since this is in the encoding path and I think we encode with a const ref do we need to copy here?

Reply:
StreamDecoderFilterCallbacks::encodeHeaders takes a HeaderMapPtr&&, so I don't think I have any alternatives. It seems fine though: we're going to have to create a HeaderMap for the cache entry; handing its ownership to the callbacks makes sense.

Reply:
OK but then we have to needlessly copy on every cached hit response? In any case, we can deal with this later, but maybe put in a perf TODO somewhere relevant?

Reply:
I'm not seeing where the needless copy is. Cache storage implementations will generally store a serialized form of the response headers. When ready to serve an entry, they'll create a HeaderMap from it, owned by a HeaderMapPtr, which they'll give to encodeHeaders. Implementations might copy the headers into the HeaderMap, or they might create a new subclass that references the serialized form. If an implementation chooses zero-copy, these interfaces won't get in the way.

Reply:
It seems like an in-memory implementation could store the actual HeaderMap object. In any case, we can deal with this later.